### ISSUE TITLE
Update identifier

### ISSUE THREAD

**Issue by alanpoulain** [doctrine/mongodb-odm#1877]
In API Platform (https://github.com/api-platform/core), we would like to allow an update of the identifier.
However it is not possible in mongodb-odm.
For instance, given this id of a document:
```php
 /**
 * @ODM\Id(strategy="NONE", type="string")
 */
private $slug;
```
It's not possible to update it because of:
https://github.com/doctrine/mongodb-odm/blob/da5074b8b54f0e4c0d342d15c56dd054d4a136a7/lib/Doctrine/ODM/MongoDB/Persisters/PersistenceBuilder.php#L121-L124
(first related commit: https://github.com/doctrine/mongodb-odm/commit/ed789d7e7d5a75d41b8d49e04a4c65bd97343a96)
I think it should be possible if the document is replaced by another with the modified identifier.
WDYT?

**Comment by malarzm** [comment]
Can't do, [`_id` is immutable](https://docs.mongodb.com/manual/core/document/#field-names) in MongoDB. Offhand, updating identifier doesn't sound like a wise thing to do :)

> The field name _id is reserved for use as a primary key; its value must be unique in the collection, is immutable, and may be of any type other than an array.

**Comment by alanpoulain** [comment]
That's why I said not updating the `_id` but replacing the targeted document with another one.
It's possible in Doctrine ORM though and it can be really useful like the slug example I wrote.

**Comment by malarzm** [comment]
Current commit flow is orphan removals, upserts, inserts, updates and lastly deletions. If we were to insert new document and later delete "old" one this can break if one would have an unique index on some other field. If we'd switch removals to be first and something would break before insertion (i.e. other operation) then data would be lost. Even if we'd work this around with transactions, there could be invalid references left in the db, as other documents would still have ref to `original-slug`, not to `updated-slug` and ODM can't track all references to a document.

All in all I believe there are way too many things that can go south and user should do such non-trivial operation on his/her own. Or treat ids as immutable, as MongoDB suggests. Reopening, but I'm still 👎 on this. @alcaeus @jmikola thoughts?

**Comment by alcaeus** [comment]
I agree with @malarzm here, this sounds like a lot of complexity for nothing. I don't quite understand why you'd want to persist the same document with a different identifier? At this time, I don't think this is something we're going to implement here.

**Comment by malarzm** [comment]
IMO using user-modifiable data as an identifier is asking for trouble and shouldn't be advocated. If something seems like an id provided by user (be it slug or email as per your examples), put an unique index on the field and use auto-generated value for the actual identifier (be it MongoId or UUID in case of ODM). The auto-generated id can be hidden from user/developer but will be there for references or can be more performant for the storage. The only cost of this is writing a method to fetch document/entity basing on "user provider id" and using that instead of `$dm->find()`/`$repository->find()` calls.
